kernel/init.c

```diff
@@ -477,6 +477,13 @@ FUNC_NORETURN void z_cstart(void)
 
 #ifdef CONFIG_MULTITHREADING
 	switch_to_main_thread(prepare_multithreading());
+#else
+#ifdef ARCH_SWITCH_TO_MAIN_NO_MULTITHREADING
+	/* Custom ARCH-specific routine to switch to main()
+	 * in the case of no multi-threading.
+	 */
+	arch_switch_to_main_no_multithreading(bg_thread_main,
+		NULL, NULL, NULL);
 #else
 	bg_thread_main(NULL, NULL, NULL);
```

Comment:
Shouldn't the `#ifdef ARCH_SWITCH_TO_MAIN_NO_MULTITHREADING` end here, so that the execution falls into the `while (true)` loop below in case `bg_thread_main()` returns? See the end of `bg_thread_main()`:
https://github.com/zephyrproject-rtos/zephyr/blob/b8f4f2032c862979b6c0eaa28e008f1c998fb904/kernel/init.c#L261-L270.

Reply:
I think this is complicated to do and might not worth the effort.
Normally, if a function changes the stack pointer, it should not return to the caller (using the caller convention). Such functions should not return. That's what happens, practically in `switch_to_main_thread(prepare_multithreading());`
Might be ok if the caller does not use the stack, but it feels _too_ error prone.

Reply:
Then maybe such infinite loop should be added to `z_arm_switch_to_main_no_multithreading()` so that this function indeed cannot return?

Reply:
Yes, that I could do. I'll mimic the init.c  and add an irq_lock and a while(1)